Makefile

```
@@ -0,0 +1,202 @@
+# Copyright (C) 2024, The Valkey contributors
+# All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+
+# Path to the code repo.
+VALKEY_ROOT ?= ../valkey
+
+# Where to install man pages
+INSTALL_MAN_DIR ?= /usr/local/share/man
+
+# Where to put the generated pages.
+BUILD_DIR ?= _build
+MD_DIR ?= $(BUILD_DIR)/md
+MAN_DIR ?= $(BUILD_DIR)/man
+HTML_DIR ?= $(BUILD_DIR)/html
+
+# -----------------------------------
+
+.PHONY: all man html md clean distclean install uninstall
+
+all: man
+
+# ---- Sanity check ----
+
+ifeq ("$(wildcard $(VALKEY_ROOT))","")
+    $(error Please provide the VALKEY_ROOT variable pointing to the Valkey source code)
+endif
+
+ifeq ("$(shell which pandoc)","")
+    $(error Please install pandoc)
+endif
+
+# ---- Source files ----
+
+topics   = $(wildcard topics/*)
+commands = $(wildcard commands/*.md)
+
+# ---- Temp files ----
+
+# JSON files for the commands that have a .md file (excluding undocumented commands).
+json_for_documented_commands = $(commands:commands/%.md=$(VALKEY_ROOT)/src/commands/%.json)
+
+$(BUILD_DIR)/.commands-per-group.json: $(VALKEY_ROOT)/src/commands/. utils/build-command-groups.py | $(BUILD_DIR)
+	utils/build-command-groups.py $(json_for_documented_commands) > $@
+$(BUILD_DIR):
+	mkdir -p $@
+
+# ---- Pre-processed markdown files (make md) ----
+
+md_topics_dir = $(MD_DIR)/topics
+md_topics = $(topics:topics/%=$(md_topics_dir)/%)
+$(md_topics): | $(md_topics_dir)
+
+md_commands_dir  = $(MD_DIR)/commands
+md_commands      = $(commands:commands/%=$(md_commands_dir)/%)
+$(md_commands): | $(md_commands_dir)
+
+md_targets = $(md_topics) $(md_commands) $(md_commands_dir)/index.md
+
+$(md_topics_dir) $(md_commands_dir):
+	mkdir -p $@
+
+$(md_topics_dir)/%.md: topics/%.md utils/preprocess-markdown.py | $(md_topics_dir)
+	utils/preprocess-markdown.py $< > $@
+$(md_topics_dir)/%.png: topics/%.png | $(md_topics_dir)
+	cp $< $@
+$(md_topics_dir)/%.gif: topics/%.gif | $(md_topics_dir)
+	cp $< $@
+
+$(md_commands_dir)/index.md: $(BUILD_DIR)/.commands-per-group.json groups.json utils/build-command-index.py | $(md_commands_dir)
+	utils/build-command-index.py --suffix .md \
+	 --groups-json groups.json \
+	 --commands-per-group-json $(BUILD_DIR)/.commands-per-group.json > $@
+
+$(md_commands_dir)/%.md: commands/%.md $(VALKEY_ROOT)/src/commands/%.json $(BUILD_DIR)/.commands-per-group.json \
+                         utils/preprocess-markdown.py utils/command_syntax.py
+	utils/preprocess-markdown.py --page-type command \
+	 --commands-per-group-json $(BUILD_DIR)/.commands-per-group.json \
+	 --valkey-root $(VALKEY_ROOT) $< > $@
+
+# ---- HTML (make html) ----
+
+html_topics_dir = $(HTML_DIR)/topics
+html_topics     = $(topics:topics/%.md=$(html_topics_dir)/%.html)
+html_topics_pics = $(topics:topics/%.png=$(html_topics_dir)/%.png) $(topics:topics/%.gif=$(html_topics_dir)/%.gif)
```

```suggestion
html_topics_dir = $(HTML_DIR)/topics
html_topics     = $(filter $(html_topics_dir)/%.html, $(topics:topics/%.md=$(html_topics_dir)/%.html))
html_topics_pics = $(filter $(html_topics_dir)/%.png, $(topics:topics/%.png=$(html_topics_dir)/%.png)) $(filter $(html_topics_dir)/%.gif, $(topics:topics/%.gif=$(html_topics_dir)/%.gif))
```

This isn't working the way I thought it was. When doing the string substitution, it's leaving everything that doesn't match as is, which are in the current directory. So when we hit clean later, it was cleaning them all up.

---

Yeah, and `topics/*.{png,gif}` are in man_targets as well. Thanks for testing and finding it!